---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "",
  fig.path = "README/README-",
  fig.align = 'center',
  fig.retina=2, 
  warning = FALSE,
  message = FALSE
)
options("width"=200)
```

iNEXT.beta3D (R package)
=====

<h5 align="right">Latest version: `r Sys.Date()`</h5>

<font color="394CAE"><h3 color= 394CAE style = "font-weight: bold"> Introduction to iNEXT.beta3D (R package): Excerpt from iNEXT.beta3D User’s Guide </h3> </font>
<br>
<h5><b>Anne Chao, Kai-Hsiang Hu, Y.S. Kao and Z.C. Szu-tu</b>
<br><br>
<i>Institute of Statistics, National Tsing Hua University, Hsin-Chu, Taiwan 30043</i>  
</h5>
<br>
iNEXT.beta3D (iNterpolation and EXTrapolation for three dimensions of Beta diversity) is an original R package available in [Github](https://github.com/AnneChao). 
An online version of [iNEXT.beta3D Online](https://chao.shinyapps.io/iNEXT_beta3D/) is also available for users without an R background. Detailed information about all functions in iNEXT.beta3D is provided in the iNEXT.beta3D Manual in [iNEXT.beta3D_vignettes](http://chao.stat.nthu.edu.tw/wordpress/wp-content/uploads/software/A%20Quick%20Introduction%20to%20iNEXT.beta3D.html), which is available from [Anne Chao's website](http://chao.stat.nthu.edu.tw/wordpress/software_download/).

`iNEXT.beta3D` focuses on Chiu's et al.'s multiple assemblages decomposition (Chiu et al., 2014) based on three measures of Hill numbers of order q: species richness (`q = 0`), Shannon diversity (`q = 1`, the exponential of Shannon entropy) and Simpson diversity (`q = 2`, the inverse of Simpson concentration). Besides, `iNEXT.beta3D` also promote phylogenetic diversity level and functional diversity level (Chao et al., 2019). For each diversity measure, `iNEXT.beta3D` uses the observed sample of abundance or incidence data (called the "reference sample") to compute diversity estimates and the associated 95% confidence intervals for the following two types of rarefaction and extrapolation (R/E): 

1. Sample-size-based (or size-based) R/E sampling curves: `iNEXT.beta3D` computes <font color=#FF6781>gamma and alpha</font> diversity estimates for rarefied and extrapolated samples up to an appropriate size. This type of sampling curve plots the diversity estimates with respect to sample size.     

2. Sample-coverage-based (or coverage-based) R/E sampling curves: `iNEXT.beta3D` computes <font color=#FF6781>gamma, alpha, and beta</font> diversity estimates and <font color=#FF6781>four dissimilarity</font> measures estimates for rarefied and extrapolated samples based on a standardized level of sample completeness (as measured by sample coverage) up to an appropriate coverage value. This type of sampling curve plots the diversity estimates with respect to sample coverage. 

`iNEXT.beta3D` also plots the above two types of sampling curves for gamma, alpha, and beta diversity and four dissimilarity-turnover.


### SOFTWARE NEEDED TO RUN iNEXT.beta3D IN R

-   Required: [R](http://cran.rstudio.com/)
-   Suggested: [RStudio IDE](http://www.rstudio.com/ide/download/)

## HOW TO RUN iNEXT.beta3D:
The `iNEXT.beta3D` package will be submitted to [CRAN] and can be downloaded from Anne Chao's [iNEXT.beta3D_github](https://github.com/AnneChao/iNEXT.beta3D) using the following commands. For a first-time installation, an additional visualization extension package (`ggplot2`) and (`iNEXT.3D`) from Anne Chao's github must be installed and loaded. 

```{r install, eval=FALSE}
# install_github('AnneChao/iNEXT.3D')
# library(iNEXT.3D)

## install iNEXT.beta3D package from CRAN
# install.packages("iNEXT.beta3D")  # coming soon

## install the latest version from github
install.packages('devtools')
library(devtools)
install_github('AnneChao/iNEXT.beta3D')

## import packages
library(iNEXT.beta3D)
```


Here are two main functions we provide in this package : 

- **iNEXTbeta3D** : Computing standardized gamma, alpha, beta diversity, and four dissimilarity-turnover indices for three dimensions: taxonomic, phylogenetic and functional diversity at specified sample coverage or sample size.

- **ggiNEXTbeta3D** : Visualizing the output from the function `iNEXTbeta3D`.


### MAIN FUNCTION: iNEXTbeta3D()

We first describe the main function `iNEXTbeta3D()` with default arguments: 

```{r eval=FALSE}
iNEXTbeta3D(data, diversity = "TD", q = c(0, 1, 2), datatype = "abundance",
            base = "coverage", level = NULL, nboot = 20, conf = 0.95,
            PDtree = NULL, PDreftime = NULL, PDtype = "meanPD",
            FDdistM = NULL, FDtype = "AUC", FDtau = NULL, FDcut_number = 30)
```

The arguments of this function are briefly described below, and will be explained in more details by illustrative examples in later text. This main function computes gamma, alpha and beta diversity estimates of order q at specified sample coverage or sample size and measures of dissimilarity. <font color=#FF6781>By default of</font> <code>base = "coverage"</code> and <code>level = NULL</code>, then this function computes the gamma, alpha, beta diversity, and four dissimilarity-turnover indices estimates up to one (for q = 1, 2) or up to the coverage of double the reference sample size (for q = 0). <font color=#FF6781>If</font> <code>base = "size"</code> and <code>level = NULL</code>, then this function computes the gamma and alpha diversity estimates up to double the reference sample size in each region. 

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(comment="", message=FALSE, warning=FALSE)
library(iNEXT.beta3D)
library(ggplot2)
```

```{r, echo=FALSE,warning=FALSE}
Des <- c("data","(a) For datatype = 'abundance', data can be input as a matrix/data.frame (species by assemblages), or a list of matrices/data.frames, each matrix represents species-by-assemblages abundance matrix.
(b) For datatype = 'incidence_raw' data can be input as a list (a region) with several lists (assemblages) of matrices/data.frames, each matrix represents species-by-sampling units.",
"diversity"," selection of diversity type: 'TD' = Taxonomic diversity, 'PD' = Phylogenetic diversity, and 'FD' = Functional diversity.",
"q","	a numerical vector specifying the diversity orders. Default is c(0, 1, 2).",
"datatype","data type of input data: individual-based abundance data (datatype = 'abundance'), sampling-unit-based incidence frequencies data (datatype = 'incidence_freq'), or species by sampling-units incidence matrix (datatype = 'incidence_raw') with all entries being 0 (non-detection) or 1 (detection).", 
"base","Sample-sized-based rarefaction and extrapolation for gamma and alpha diversity (base = 'size') or coverage-based rarefaction and extrapolation for gamma, alpha and beta diversity (base = 'coverage'). Default is base = 'coverage'.",
"level","A numerical vector specifying the particular value of sample coverage (between 0 and 1 when base = 'coverage') or sample size (base = 'size'). level = 1 (base = 'coverage') means complete coverage(the corresponding diversity represents asymptotic diversity). If base = 'size' and 'level = NULL', then this function computes the gamma and alpha diversity estimates up to double the reference sample size.
If base = 'coverage' and 'level = NULL', then this function computes the gamma and alpha diversity estimates up to one (for 'q = 1, 2') or up to the coverage of double the reference sample size (for 'q = 0'); the corresponding beta diversity is computed up to the same maximum coverage as the alpha diversity.",
"nboot"," a positive integer specifying the number of bootstrap replications when assessing sampling uncertainty and constructing confidence intervals. Enter 0 to skip the bootstrap procedures. Default is 50.",
"conf","a positive number < 1 specifying the level of confidence interval. Default is '0.95'.",
"PDtree","	(required only when diversity = 'PD'), a phylogenetic tree in Newick format for all observed species in the pooled assemblage.",
"PDreftime", "(required only when diversity = 'PD'), a vector of numerical values specifying reference times for PD. Default is NULL (i.e., the age of the root of PDtree).",
"PDtype"," (required only when diversity = 'PD'), select PD type: PDtype = 'PD' (effective total branch length) or PDtype = 'meanPD' (effective number of equally divergent lineages). Default is 'meanPD', where meanPD = PD/tree depth.",
"FDdistM"," (required only when diversity = 'FD'), select FD type: FDtype = 'tau_values' for FD under specified threshold values, or FDtype = 'AUC' (area under the curve of tau-profile) for an overall FD which integrates all threshold values between zero and one. Default is 'AUC'.",
"FDtype"," (required only when diversity = 'FD'), select FD type: FDtype = 'tau_values' for FD under specified threshold values, or FDtype = 'AUC' (area under the curve of tau-profile) for an overall FD which integrates all threshold values between zero and one. Default is 'AUC'.",
"FDtau"," (required only when diversity = 'FD' and FDtype = 'tau_values'), a numerical vector between 0 and 1 specifying tau values (threshold levels). If NULL (default), then threshold is set to be the mean distance between any two individuals randomly selected from the pooled assemblage (i.e., quadratic entropy).",
"details"," a logical variable to decide whether do you want to print out the detailed value for each plots, default is `FALSE`.",
"FDcut_number","(required only when diversity = 'FD' and FDtype = 'AUC'), a numeric number to split zero to one into several equal-spaced length. Default is '30'.")
output <- 
  matrix(Des, 
         ncol=2, byrow = TRUE)

library(htmlTable)
htmlTable(output,
          header =  c("Argument","Description"),align = "l"
          )
```

This function returns an `"iNEXTbeta3D"` object which can be further used to make plots using the function `ggiNEXTbeta3D()` to be described below. 


## DATA FORMAT/INFORMATION

Two types of data are supported:

1. Individual-based abundance data (`datatype = "abundance"`):
Input data for each region with several assemblages/sites include samples species abundances in an empirical sample of n individuals (“reference sample"). When there are N assemblages in a region, input data consist of a list with an S by N abundance matrix; For M regions consisting N assemblages, input data should be <font color=#FF6781>a list of M S by N abundance matrices</font>.

2. Sampling-unit-based incidence raw data (`datatype = "incidence_raw"`):
Input data for a reference sample consist of a species-by-sampling-unit matrix; The number of sampling units and species for each assemblages in a region should be the same.
For M regions consisting N assemblages, input data should be 
<font color=#FF6781>a list of M lists each consists of N species-by-sampling-unit matrices</font>.


Two data sets ( Brazil rain-forests under different habitats for abundance data and Beetles under different time periods for incidence raw data ) are included in `iNEXT.beta3D` package for illustration. For abundance data, the Brazil rain-forests data consist a list of twelve matrices (here we use two sites "Marim" and "Edge" for computing). Each matrix represents a species-by-habitat ("Edge" and "Interior") in a site. For incidence raw data, the beetles data consist two lists ("unlogged" and "logged"). For each list, it has two matrices (each means year 2008 and 2009) which represent species-by-sampling units assemblages. For these data, the following commands display how to compute taxonomic, phylogenetic and functional beta decomposition and dissimilarities at specified sample coverage for abundance and incidence raw data respectively.

Run the following code to view Brazil rain-forests abundance data: (Here we only show the first ten rows for each matrix)

```{r eval=FALSE}
data("Brazil_data")
Brazil_data
```

```{r echo=FALSE}
data("Brazil_data")
lapply(Brazil_data, function(x) x[1:10,])
```

Run the following code to view Beetles incidence raw data: (Here we only show the first ten rows and six columns for each matrix)

```{r eval=FALSE}
data("beetle_inc")
beetle_abu
```

```{r echo=FALSE}
data("beetle_inc")
lapply(beetle_inc, function(x) lapply(x, function(y) y[1:10,1:6]))
```


## Rarefaction/Extrapolation Via Examples

#### Output of iNEXTbeta3D()

When `base = 'coverage'`, the `iNEXTbeta3D()` function returns the `"iNEXTbeta3D"` object including seven data frames for each regions: 

- gamma ( gamma diversity )
- alpha ( alpha diversity )
- beta ( beta diversity )
- C ( Sorensen-type non-overlap )
- U ( Jaccard-type  non-overlap )
- V ( Sorensen-type turnover )
- S ( Jaccard-type  turnover )

When `base = 'size'`, the `iNEXTbeta3D()` function returns the `"iNEXTbeta3D"` object including two data frames for each regions: 

- gamma ( gamma diversity )
- alpha ( alpha diversity )


## Taxonomic Diversity ##

First, we run the `iNEXTbeta3D()` function with Brazil rain-forests abundance data to compute standardized gamma, alpha, beta diversity, and four dissimilarities under `base = 'coverage'`. (Here we only show the first six outputs for each output in region "Marim")

```{r eval=FALSE}
data("Brazil_data")

# Taxonomic diversity
Abundance_TD = iNEXTbeta3D(data = Brazil_data[c('Marim', 'Rebio 2')], diversity = 'TD', datatype = "abundance", 
                           base = 'coverage', nboot = 100)
Abundance_TD
```


```{r echo=FALSE}
Abundance_TD = iNEXTbeta3D(data = Brazil_data[c('Marim', 'Rebio 2')], diversity = 'TD', datatype = "abundance", 
                           base = 'coverage', nboot = 100)
lapply(Abundance_TD$Marim, function(x) {
    tmp = x[1:6,]
    tmp[,c('Estimate', 'Size', 's.e.', 'LCL', 'UCL')] = round(tmp[,c('Estimate', 'Size', 's.e.', 'LCL', 'UCL')], 3)
    tmp
})
```

The output contains seven data frames: `gamma`, `alpha`, `beta`, `C`, `U`, `V`, `S`. For each data frame, it includes the diversity estimate (`Estimate`), the diversity order (`Order.q`), `Method` (Interpolated, Observed, or Extrapolated, depending on whether the size `m` is less than, equal to, or greater than the reference sample size), the sample coverage estimate (`SC`), the sample size (`Size`), the standard error from bootstrap replications (`s.e.`), the 95% lower and upper confidence limits of diversity (`LCL`, `UCL`), and the name of region (`Region`). These diversity estimates with confidence intervals are used for plotting the diversity curve.


Then, we can also use beetles incidence raw data to compute standardized gamma, alpha, beta diversity, and four dissimilarities under `base = 'coverage'`. (Here we don't show the output)

```{r eval=FALSE}
data("beetle_inc")

Incidence_TD = iNEXTbeta3D(data = beetle_inc, diversity = 'TD', datatype = "incidence_raw",
                           base = 'coverage', nboot = 30)
Incidence_TD
```


## Phylogenetic Diversity ##

We can compute standardized gamma, alpha, beta diversity under phylogenetic level with newick format data (`beetle_tree`). Here, `PDtree` is the essential setting, and user can select `PDtype` and `PDreftime` alternatively. `PDtype` has two selection: `"PD"` (effective total branch length) or `"meanPD"` (effective number of equally divergent lineages). `PDreftime` means a numerical value specifying reference time for computing phylogenetic diversity. If `PDreftime = NULL`, then it is set the reference time to the age of the root of PDtree. (Here we don't show the output)

```{r eval=FALSE}
## Phylogenetic diversity
# Abundance data
data("Brazil_PDFD_data")
data("Brazil_tree")

Abundance_PD = iNEXTbeta3D(data = Brazil_PDFD_data[c('Marim', 'Rebio 2')], diversity = 'PD', datatype = "abundance", 
                           base = 'coverage', nboot = 10, PDtree = Brazil_tree)
Abundance_PD


# Incidence data
data("beetle_inc")
data("beetle_tree")

Incidence_PD = iNEXTbeta3D(data = beetle_inc, diversity = 'PD', datatype = "incidence_raw", 
                           base = 'coverage', nboot = 30, PDtree = beetle_tree)
Incidence_PD
```

## Functional diversity 

We can compute standardized gamma, alpha, beta diversity under functional level with distance matrix (`beetle_distM`). Here, `FDdistM` is the essential setting, and user can select `FDtype`, `FDtau`, and `FDcut_number` alternatively. `FDtype` has two selections: "tau_value" means compute functional diversity under a specific threshold value and "AUC" means compute functional diversity by integrating all threshold values between zero and one. If user selects `FDtype = "tau_value"`, it need to input a `FD_tau` threshold value. If `FD_tau = NULL`, then it is set to be the mean distance between any two individuals randomly selected from the pooled assemblage (i.e., quadratic entropy). (Here we don't show the output)

```{r eval=FALSE}
## Functional diversity - under specified threshold
# Abundance data
data("Brazil_PDFD_data")
data("Brazil_distM")

Abundance_FDtau = iNEXTbeta3D(data = Brazil_PDFD_data[c('Marim', 'Rebio 2')], diversity = 'FD', 
                              datatype = "abundance", base = 'coverage', nboot = 30, 
                              FDdistM = Brazil_distM, FDtype = 'tau_value')
Abundance_FDtau


# Incidence data
data("beetle_inc")
data("beetle_distM")

Incidence_FDtau = iNEXTbeta3D(data = beetle_inc, diversity = 'FD', datatype = "incidence_raw", 
                              base = 'coverage', nboot = 30, 
                              FDdistM = beetle_distM, FDtype = 'tau_value')
Incidence_FDtau



## Functional diversity - area under curve (integrates all threshold values between zero and one)
# Abundance data
Abundance_FDAUC = iNEXTbeta3D(data = Brazil_PDFD_data[c('Marim', 'Rebio 2')], diversity = 'FD', 
                              datatype = "abundance", base = 'coverage', nboot = 30, 
                              FDdistM = Brazil_distM, FDtype = 'AUC')
Abundance_FDAUC


# Incidence data
Incidence_FDAUC = iNEXTbeta3D(data = beetle_inc, diversity = 'FD', datatype = "incidence_raw", 
                              base = 'coverage', nboot = 30, 
                              FDdistM = beetle_distM, FDtype = 'AUC')
Incidence_FDAUC
```


## GRAPHIC DISPLAYS: FUNCTION ggiNEXTbeta3D()

The function `ggiNEXTbeta3D()`, which extends `ggplot2` to the `"iNEXTbeta3D"` object with default arguments, is described as follows: 

```{r eval=FALSE}
ggiNEXTbeta3D(output, type = "B", scale = "free", transp = 0.4)  
```


```{r, echo=FALSE,warning=FALSE}
Des <- c("output","",
"type","(required only when base = 'coverage'), selection of plot type : type = 'B' for plotting the gamma, alpha, and beta diversity ; type = 'D' for plotting 4 turnover dissimilarities.",
"scale","Are scales shared across all facets (the default, 'fixed'), or do they vary across rows ('free_x'), columns ('free_y'), or both rows and columns ('free')? Default is 'free'.",
"transp","a value between 0 and 1 controlling transparency. 'transp = 0' is completely transparent, default is '0.4'.")
output <- 
  matrix(Des, 
         ncol=2, byrow = TRUE)

library(htmlTable)
htmlTable(output,
          header =  c("Argument","Description"),align = "l"
          )
```

The `ggiNEXTbeta3D()` function is a wrapper around the `ggplot2` package to create a R/E curve using a single line of code. The resulting object is of class `"ggplot"`, so it can be manipulated using the `ggplot2` tools. Users can visualize the output of beta diversity or four dissimilarities by setting the parameter <code>**type**</code>:

```{r, fig.align='center', fig.height=6, fig.width=6}
Abundance_TD = iNEXTbeta3D(data = Brazil_data[c('Marim', 'Rebio 2')], diversity = 'TD', datatype = 'abundance', 
                           base = "coverage", nboot = 30, conf = 0.95)
ggiNEXTbeta3D(Abundance_TD, type = 'B')
```


```{r, fig.align='center', fig.height=8, fig.width=6}
ggiNEXTbeta3D(Abundance_TD, type = 'D')
```


The following commands return the size-based R/E sampling curves for gamma and alpha taxonomic diversity:

```{r, fig.align='center', fig.height=4, fig.width=6}
output = iNEXTbeta3D(data = Brazil_data[c('Marim', 'Rebio 2')], diversity = 'TD', datatype = 'abundance', 
                     base = "size", nboot = 30, conf = 0.95)
ggiNEXTbeta3D(output)
```



## HOW TO CITE iNEXT.beta3D
If you publish your work based on the results from the `iNEXT.beta3D` package, you should make references to the following methodology paper:

- Chao, A., Thorn, S., Chiu, C.-H., Moyes, F., Hu, K.-H., Chazdon, R. L., Wu, J., Dornelas, M., Zelen??, D., Colwell, R. K., and Magurran, A. E. (2023). Rarefaction and extrapolation with beta diversity under a framework of Hill numbers: the iNEXT.beta3D standardization. <i>Ecological Monographs</i>, e1588.


## License
The iNEXT.beta3D package is licensed under the GPLv3. To help refine `iNEXT.beta3D`, your comments or feedback would be welcome (please send them to Anne Chao or report an issue on the iNEXT.beta3D github [iNEXT.beta3D_github](https://github.com/AnneChao/iNEXT.beta3D). 

## References
- Chiu, C.-H., Jost, L. and Chao*, A. (2014). Phylogenetic beta diversity, similarity, and differentiation measures based on Hill numbers. Ecological Monographs 84, 21-44.

- Chao, A., Chiu, C.-H., Vill??ger, S., Sun, I.-F., Thorn, S., Lin, Y.-C., Chiang, J. M. and Sherwin, W. B. (2019). An attribute-diversity approach to functional diversity, functional beta diversity, and related (dis)similarity measures. Ecological Monographs, 89, e01343. 10.1002/ecm.1343.

- Chao, A., Thorn, S., Chiu, C.-H., Moyes, F., Hu, K.-H., Chazdon, R. L., Wu, J., Dornelas, M., Zelen??, D., Colwell, R. K., and Magurran, A. E. (2023). Rarefaction and extrapolation with beta diversity under a framework of Hill numbers: the iNEXT.beta3D standardization. <i>Ecological Monographs</i>, e1588.

